File: src/Cocoar.SignalARRR.Swift/Sources/CocoarSignalARRR/HARRRConnectionOptions.swift
import Foundation

/// A credential: called whenever the credential is needed, so a refreshed token is picked up.
public typealias HARRRCredential = @Sendable () async -> String

/// Configuration options for `HARRRConnection`.
///
/// A connection has two credentials, checked by different things. The **connection credential**
/// authenticates negotiate and the transport and is what `[Authorize]` on the hub class checks. The
/// **message credential** travels with every message, answers token challenges and authorises file
/// transfers, and is what `[Authorize]` on a method or a `ServerMethods` class checks. The options
/// are named the same in every SignalARRR client, and nothing is coupled implicitly: setting one of
/// the two sets only that one.
public struct HARRRConnectionOptions: Sendable {

    /// One credential for the connection and for every message — the common case. A
    /// `connectionCredential` or `messageCredential` set alongside it takes its part over.
    public var credential: HARRRCredential?

    /// Authenticates the connection: the negotiate request and the transport. Called on every
    /// connect and reconnect. Only available through `HARRRConnection.create(url:)`; a
    /// `SignalRWebSocketClient` that is already built has its own.
    public var connectionCredential: HARRRCredential?

    /// Authenticates each message: travels as `ClientRequestMessage.Authorization`, answers token
    /// challenges, and authorises file-transfer requests. Called per use.
    public var messageCredential: HARRRCredential?

    public init(
        credential: HARRRCredential? = nil,
        connectionCredential: HARRRCredential? = nil,
        messageCredential: HARRRCredential? = nil
    ) {
        self.credential = credential
        self.connectionCredential = connectionCredential
        self.messageCredential = messageCredential
    }

    var usesCredentials: Bool {
        credential != nil || connectionCredential != nil || messageCredential != nil
    }

    var resolvedConnectionCredential: HARRRCredential? { connectionCredential ?? credential }
    var resolvedMessageCredential: HARRRCredential? { messageCredential ?? credential }
}

/// A connection was configured in a way that cannot work, typically a credential set in two places.
/// `HARRRConnection.create` does not throw, so this surfaces from `start()`.
public struct HARRRConfigurationError: Error, LocalizedError, CustomStringConvertible {
    public let message: String
    public var errorDescription: String? { message }
    public var description: String { message }
}

File: src/Cocoar.SignalARRR.Client.FullFramework/HARRRConnectionOptions.cs
using System;
using System.Threading.Tasks;

namespace Cocoar.SignalARRR.Client.FullFramework {

    /// <summary>
    /// SignalARRR's options for a connection — the same credential options as every other SignalARRR
    /// client. The <b>connection credential</b> authenticates negotiate and the transport and is what
    /// <c>[Authorize]</c> on the hub class checks; the <b>message credential</b> travels with every
    /// message, answers authentication challenges and authenticates file transfers, and is what
    /// <c>[Authorize]</c> on a method or a <c>ServerMethods</c> class checks. Nothing is coupled
    /// implicitly: setting one of the two sets only that one.
    /// </summary>
    public class HARRRConnectionOptions {

        /// <summary>
        /// Authenticates the connection: the negotiate request and the transport. Set on SignalR's
        /// <c>AccessTokenProvider</c>, so it is called on every connect. Only available through
        /// <c>HARRRConnection.Create(builder => ..., options => ...)</c>; a <c>HubConnection</c> that is
        /// already built has its own.
        /// </summary>
        public Func<Task<string>> ConnectionCredential { get; set; }

        /// <summary>
        /// The credential SignalARRR sends with every message, as the answer to an authentication
        /// challenge, and on file-transfer requests. Called per use, so a refreshed token is picked up.
        /// </summary>
        public Func<Task<string>> MessageCredential { get; set; }
    }

    public class HARRRConnectionOptionsBuilder {

        private HARRRConnectionOptions Options { get; } = new HARRRConnectionOptions();

        public static implicit operator HARRRConnectionOptions(HARRRConnectionOptionsBuilder builder) {
            return builder?.Options;
        }

        /// <summary>
        /// One credential for the connection and for every message — the common case. Resolved per
        /// use, for a token that expires or is refreshed.
        /// </summary>
        public HARRRConnectionOptionsBuilder WithCredential(Func<Task<string>> credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            Options.ConnectionCredential = credential;
            Options.MessageCredential = credential;
            return this;
        }

        /// <summary>One credential for the connection and for every message, resolved synchronously.</summary>
        public HARRRConnectionOptionsBuilder WithCredential(Func<string> credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithCredential(() => Task.FromResult(credential()));
        }

        /// <summary>One credential for the connection and for every message that does not change.</summary>
        public HARRRConnectionOptionsBuilder WithCredential(string credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithCredential(() => Task.FromResult(credential));
        }

        /// <summary>
        /// Authenticates the connection only — negotiate and the transport, what <c>[Authorize]</c>
        /// on the hub class checks. Called on every connect.
        /// </summary>
        public HARRRConnectionOptionsBuilder WithConnectionCredential(Func<Task<string>> credential) {
            Options.ConnectionCredential = credential ?? throw new ArgumentNullException(nameof(credential));
            return this;
        }

        /// <summary>Authenticates the connection only, resolved synchronously.</summary>
        public HARRRConnectionOptionsBuilder WithConnectionCredential(Func<string> credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithConnectionCredential(() => Task.FromResult(credential()));
        }

        /// <summary>Authenticates the connection only, with a credential that does not change.</summary>
        public HARRRConnectionOptionsBuilder WithConnectionCredential(string credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithConnectionCredential(() => Task.FromResult(credential));
        }

        /// <summary>
        /// Authenticates every message only — what <c>[Authorize]</c> on a method or a
        /// <c>ServerMethods</c> class checks, the answer to a challenge, and file transfers. Resolved
        /// per use.
        /// </summary>
        public HARRRConnectionOptionsBuilder WithMessageCredential(Func<Task<string>> credential) {
            Options.MessageCredential = credential ?? throw new ArgumentNullException(nameof(credential));
            return this;
        }

        /// <summary>Authenticates every message only, resolved synchronously.</summary>
        public HARRRConnectionOptionsBuilder WithMessageCredential(Func<string> credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithMessageCredential(() => Task.FromResult(credential()));
        }

        /// <summary>Authenticates every message only, with a credential that does not change.</summary>
        public HARRRConnectionOptionsBuilder WithMessageCredential(string credential) {
            if (credential == null) throw new ArgumentNullException(nameof(credential));
            return WithMessageCredential(() => Task.FromResult(credential));
        }
    }
}
